OpenRA.Mods.Common/Warheads/DamageWarhead.cs

```diff
@@ -35,32 +35,48 @@ public override bool IsValidAgainst(Actor victim, Actor firedBy)
 			return base.IsValidAgainst(victim, firedBy);
 		}
 
-		public int DamageVersus(Actor victim)
+		public int DamageVersus(Actor victim, HitShapeInfo shapeInfo)
 		{
 			var armor = victim.TraitsImplementing<Armor>()
-				.Where(a => !a.IsTraitDisabled && a.Info.Type != null && Versus.ContainsKey(a.Info.Type))
+				.Where(a => !a.IsTraitDisabled && a.Info.Type != null && Versus.ContainsKey(a.Info.Type) &&
+					(shapeInfo.ArmorTypes.Count == 0 || shapeInfo.ArmorTypes.Contains(a.Info.Type)))
 				.Select(a => Versus[a.Info.Type]);
 
 			return Util.ApplyPercentageModifiers(100, armor);
 		}
 
 		public override void DoImpact(Target target, Actor firedBy, IEnumerable<int> damageModifiers)
 		{
-			// Used by traits that damage a single actor, rather than a position
+			// Used by traits or warheads that damage a single actor, rather than a position
 			if (target.Type == TargetType.Actor)
-				DoImpact(target.Actor, firedBy, damageModifiers);
+			{
+				var victim = target.Actor;
+
+				// Cannot be damaged without a Health trait
+				if (!victim.Info.HasTraitInfo<HealthInfo>())
+					return;
+
+				if (!IsValidAgainst(victim, firedBy))
+					return;
+
+				var closestActiveShape = victim.TraitsImplementing<HitShape>().Where(Exts.IsTraitEnabled)
+					.MinByOrDefault(t => t.Info.Type.DistanceFromEdge(victim.CenterPosition, victim));
+
+				// Cannot be damaged without an active HitShape
+				if (closestActiveShape == null)
+					return;
+
+				DoImpact(victim, firedBy, closestActiveShape.Info, damageModifiers);
+			}
 			else if (target.Type != TargetType.Invalid)
 				DoImpact(target.CenterPosition, firedBy, damageModifiers);
 		}
 
 		public abstract void DoImpact(WPos pos, Actor firedBy, IEnumerable<int> damageModifiers);
 
-		public virtual void DoImpact(Actor victim, Actor firedBy, IEnumerable<int> damageModifiers)
+		public virtual void DoImpact(Actor victim, Actor firedBy, HitShapeInfo shapeInfo, IEnumerable<int> damageModifiers)
 		{
-			if (!IsValidAgainst(victim, firedBy))
```

Comment:
Removing this from here doesn't sit well with me - this is public, so anything can call it.

Considering that this is actually effectively only called from three places (`HealthPercentageWarhead` overrides both this method and the thing that calls it) my preference is to remove this completely and inline the `InflictDamage` calls.